client/lib/logmein/index.ts

```diff
@@ -0,0 +1,129 @@
+/**
+ * Internal Dependencies
+ */
+import { isEnabled } from '@automattic/calypso-config';
+import wpcom from 'calypso/lib/wp';
+
+/**
+ * Append logmein=1 query parameter to mapped domain urls we want the user to be logged in against.
+ */
+
+// Used as placeholder / default domain to detect when we're looking at a relative url
+const INVALID_URL = `https://__domain__.invalid`;
+
+export function logmeinUrl( url: string, sites: any, isWPComLoggedIn = true ): string {
+	let newurl: URL;
+
+	if ( ! isEnabled( 'logmein' ) ) {
+		return url;
+	}
+
+	try {
+		newurl = new URL( String( url ), INVALID_URL );
+	} catch ( e ) {
+		// Ignore unparseable urls
+		return url;
+	}
+
+	// Ignore and passthrough /relative/urls that have no host specified
+	if ( newurl.origin === INVALID_URL ) {
+		return url;
+	}
+
+	let allow = logmeinAllowedUrls( sites );
+	// Ignore urls not in the allow list
+	allow = allow.map( ( allowed: string ) => new URL( allowed, INVALID_URL ).hostname );
+	if ( allow.indexOf( newurl.hostname ) === -1 ) {
+		return url;
+	}
+
+	// logmein doesn't work with http.
+	newurl.protocol = 'https:';
+
+	// we're already logged into wordpress.com so we can take a shortcut
+	if ( isWPComLoggedIn ) {
+		newurl.searchParams.set( 'logmein', 'direct' );
+	} else {
+		newurl.searchParams.set( 'logmein', '1' );
+	}
+	return newurl.toString();
+}
+
```

Comment:
Can we find if a site is a mapped simple site? I don't trust this exclusion approach. May need to extend this to atomic soon though so? :shrug:

Reply:
@niranjan [mentions](https://github.com/Automattic/wp-calypso/pull/52909#discussion_r646360802) using:

```
! isJetpackSite( state, siteId ) || ! isAtomicSite( state, siteId )
```

It looks like these are covered below (`isJetpackSite` [checks](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/sites/selectors/is-jetpack-site.js#L20) `site.jetpack`, `isAtomicSite` [checks](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/selectors/is-site-automated-transfer.js) `site.options.is_automated_transfer`.

It'd be good if anyone can chime in about these other options I'm checking. I just picked things out of the site object/options that might point at products we don't cover with logmein support.

- assuming `vip` is `vip`
- assuming `jetpack` is `jetpack`
- what's the difference between `is_wpcom_atomic` and `is_automated_transfer`? 
- assuming `domain_only` is a dns only setup, ie. server not going to be wpcom
- assuming `is_redirect` is somewhat the same as `domain_only`
- assuming `is_wpcom_store` is woo and therefore atomic
- assuming `is_wpforteams_site` is p2, @josephscott do you know if p2 supports mapped domains and therefore should be covered by logmein (i.e. remove this restriction and let it get picked up by is_mapped_domain)

I tried to find context on specifics but mgs turns up a few blanks.

Reply:
Simple site is one that is not Jetpack. Atomic sites are a subset of Jetpack sites, and have the `options.is_automated_transfer` flag.

The `is_wpcom_atomic` flag is for sites that run on a newer version of Atomic infrastructure, I think you don't need to check that here.

Reply:
Are VIP sites Jetpack?

Ah so is_wpcom_atomic is subset of is_automated_transfer. Probably don't need this for now but may come in handy later if/when we support logmein on atomic.

Reply:
> Are VIP sites Jetpack?

I'm not sure as I have zero experience with VIP, but I think a VIP site can be technically anything. Simple site, Atomic site, other hosting (vipgo) with or without Jetpack...

Reply:
> ! isJetpackSite( state, siteId ) || ! isAtomicSite( state, siteId )

I'm looking at this again and realize made a mistake. The condition after the OR is not required, just  `! isJetpackSite( state, siteId )` is what I was going for to cover simple sites. 

Going to the checks below, looks like `! site.jetpack` will exclude atomic too, so I guess we don't need separate checks for atomic. `! site.options.is_domain_only` and `site.options.is_mapped_domain` make sense in addition to the jetpack check. I think these three should cover the case for VIP, WP for teams etc. since if it's a mapped domain site that is not Jetpack, then `logmein` wouldn't technically fail. 

> do you know if p2 supports mapped domains 

@lamosty Does P2 product support custom domains yet?

Reply:
> I'm looking at this again and realize made a mistake. 

I didn't notice but I understood what you meant anyway :laughing: 

> then logmein wouldn't technically fail.

The failure mode is what scares me the most about this, with the `direct` mode of logmein if you're not logged in to wordpress.com during the redirect flow you'll land on a blank white screen.

This probably needs fixing with some kind of fallback as I suspect safari ITP might break this flow at some point by treating cookies in redirect hops as third party (blocked).

Reply:
> Does P2 product support custom domains yet?

No, we don't support them just yet. At least not officially. There are a couple of them out there from early days before we ran into problems with Safari and them, but we don't support them.

Reply:
Thanks @naxoc, I pinged @lamosty and got the following: 

- Not currently supported
- Plan to support them once 3p cookie problems are sorted
- They're all simple sites

Since they're all simple sites (and I assume going to stay that way) I'm going to treat them as valid allowing them to be logmein'd if they're also mapped. This should be a noop until we start supporting mapped domains for p2. Once we do support them, logmein redirect links should just work like any other simple site.